LLL: Fix for edge case in the create macros.

benjaminion opened the pull request:
When (msize) is initially zero, the former create commands failed due to the "lll" expression overwriting the memory at 0x00. This fix reserves memory location 0x00 before calling msize. If msize is already 32 or greater it has no effect.

Test cases are included.

axic commented:
The created code (from this input `(create (revert 0 0))`) is the following for the original:
```
  mstore(0x0, msize)
  dataSize(sub_0)
  dup1
  dataOffset(sub_0)
  mload(0x0)
  codecopy
  mload(0x0)
  0x0
  create
stop

sub_0: assembly {
      0x0
      dup1
      revert
}
```

and the new one:
```
  mstore(0x0, 0xff)
  mstore(0x0, msize)
  dataSize(sub_0)
  dup1
  dataOffset(sub_0)
  mload(0x0)
  codecopy
  mload(0x0)
  0x0
  create
stop

sub_0: assembly {
      0x0
      dup1
      revert
}
```

Can you please explain the problem and solution a bit better?

axic commented on liblll/CompilerState.cpp:
```diff
@@ -54,8 +54,8 @@ void CompilerState::populateStandard()
-	"(def 'create (value code) { [0]:(msize) (create value @0 (lll code @0)) })"
-	"(def 'create (code) { [0]:(msize) (create 0 @0 (lll code @0)) })"
+	"(def 'create (value code) { [0]:{[0]:0xff (msize)} (create value @0 (lll code @0)) })"
+	"(def 'create (code) { [0]:{[0]:0xff (msize)} (create 0 @0 (lll code @0)) })"
```
I would argue it would make more sense using a different memory location.

benjaminion replied:
Consider the original code when `msize` is zero at first:

```
  mstore(0x0, msize) <-- inserts 0 in M[0] (memory location 0)
  dataSize(sub_0)
  dup1
  dataOffset(sub_0)
  mload(0x0)  <-- This is 0
  codecopy  <-- Overwrites M[0] with the first word of sub_0!!! This is a bug.
  mload(0x0) <-- Ought to be zero, but is now probably not.
  0x0
  create <-- this will produce garbage (memory access violation, probably)
stop
```

In the second case, again with msize being zero at first:

```
 mstore(0x0, 0xff)  <-- This is a hack to ensure that msize is >=32. (When msize is already >=32 this has no effect.)
  mstore(0x0, msize) <-- now msize=32 so M[0] is set to 32
  dataSize(sub_0)
  dup1
  dataOffset(sub_0)
  mload(0x0)
  codecopy <-- copies to memory starting with M[32]. This is fine.
  mload(0x0) <-- M[0] is still 32. This is good!
  0x0
  create <-- will now do the right thing.
stop
```

Admittedly, doing the initial `[0]:0xff` is a bit hacky and obscure. Really we should set M[0] to be `(min (msize) 32)`, but we have no `min` macro, and it would take a load more opcodes and a jump. So this is the best quick fix I could come up with that just fixed this edge case but had no effect in the general case.

Another option is to change the macro to 

```
(def 'create (code) { [0]:(+ 32 (msize)) (create 0 @0 (lll code @0)) })
```

But unlike my fix that affects all cases, not just when msize<32 to start with.

I hope that is clearer. Let me know if you need more info.

axic replied:
I think I shouldn't review stuff late.

Also by looking at LLL, isn't `alloc` supposed to do the above in a non-hacky way?

benjaminion replied:
> But unlike my fix that affects all cases, not just when msize<32 to start with.

And wastes a precious word of memory in the general case, with a marginal gas cost implication :-)

benjaminion replied:
> I think I shouldn't review stuff late.

No problem :-)  - this one was really tricky to hunt down.

> Also by looking at LLL, isn't alloc supposed to do the above in a non-hacky way?

I'll have a look at `alloc` in the a.m. Past my bedtime too...

axic replied:
Is there any particular reason for using `0xff`?

I'd reorder this if possible as:
```
// NOTE: memory location 0 is set in order to force msize to be at least 32 bytes
(def 'create (code) { [0]:0 [0]:(msize) (create 0 @0 (lll code @0)) })
```

benjaminion replied:
Yes, I like this - much cleaner.

No particular reason for 0xff - only to signal that something hacky was going on. But your version is an improvement.

I've updated the commit accordingly and rebased.

benjaminion committed:
LLL: Fix for edge case in the create macros.

When (msize) is initially zero the previous create commands failed due to the "lll"
expression overwriting the memory at 0x00. This fix "reserves" memory location 0x00
before calling msize. If msize is already 32 or greater it has no effect.

benjaminion committed:
LLL: Testcases for the two create built-in macros.

benjaminion commented:
Updated to reflect @axic's review and rebased.
